plugin/src/compilation/decompiler.ml

```diff
@@ -65,13 +83,29 @@ let tac_from_term env trm : tact list =
             fresh_id_in_env in_env (Id.of_string "H") env
          | Name n -> n in
        let env = push_local (Name name, t) env in
-       (Intro name) :: first_pass env b
+       Intro name :: first_pass env b
+    (* Match on well-known functions used in the proof. *)
+    | App (f, args) ->
+       (* Applying to something that *might* be in eta-expanded 
+          form, try to reduce. *)
+       if isLambda f then
+         let app_one = mkApp (f, [|Array.get args 0|]) in
+         let beta = Reduction.whd_betaiota env app_one in
+         let app = mkApp (beta, Array.sub args 1 (Array.length args - 1)) in
+         first_pass env app
+       else
+         let left = is_rewrite_l f in
+         if left || is_rewrite_r f then
+           let prf_eq = Array.get args 5 in
```

Comment:
While we don't know if this will show up yet, to protect against curried terms, perhaps we want to check the number of arguments to make sure there are enough. If there are not enough, we can always just default to the normal "apply" case.